File: lab1-11.02.21/src/shell.c
#include "./headers/fs_utils.h"
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "./str_extensions/println.h"
#include "./str_extensions/switch.h"

int main(int argc, char *argv[]) {
    if ((argc == 2 && strcmp(argv[1], "--help") == 0) || (argc == 2 && strcmp(argv[1], "-H") == 0)) 
        println("--info : for information about disks and partitions\n--explorer <partition_path> : for editing");
    else if ((argc == 2 && strcmp(argv[1], "--info") == 0) || (argc == 2 && strcmp(argv[1], "-I") == 0))
        println(get_partitions_str());
    else if ((argc == 3 && strcmp(argv[1], "--explorer") == 0) || (argc == 3 && strcmp(argv[1], "-E") == 0)) {
        FileSystem* fs = open_fs(argv[2]);
        if (fs == NULL) println("unsupported filesystem in: %s. Supported fs is HFS+ only");
        else {
            char* input_str = malloc(FILENAME_MAX * 4 + 2); // команда + 2 аргумента + запас на пробелы
            for(;;) {
                printf("> ");
                fgets(input_str, 1024, stdin);
                char* sep = " \n";
                char* cmd = strtok(input_str, sep);
                if (cmd == NULL) continue;
                char* path = strtok(NULL, sep);
                char* cmd_arg = strtok(NULL, sep);
                switchs(cmd) {
                    cases("exit")
                        close_fs(fs);
                        return 0;
                    cases("ls")
                        println(ls(fs, path));
                        break;
                    cases("pwd")
                        println(pwd(fs));
                        break;
                    cases("cd")
                        println(cd(fs, path));
                        break;
                    cases("cp")
                        println(cp(fs, path, cmd_arg));
                        break;
                    defaults
                        println("supported commands:\npwd : show path in mounted fs;\nls <path> : show items in path\ncd <path> : moves into mounted fs\ncp <src> <dest> : copy file or folder inside mounted fs or from mounted os to external (only absolute path) supports\nexit : terminate session");
                        break;
                } switchs_end;
            }
        }
    } else println("error, mode don't specified, start with -H");
    return 0;
}

File: lab2-02.06.21/shell.js
const native = require('bindings')('spo_wrapper')
const reader = require("readline-sync")
const args = process.argv
if (args.length == 3 && ['--help', '-H'].includes(args[2])) console.log(`
    --info : for information about disks and partitions
    --explorer <partitions_path> : for editing
    `)
else if (args.length == 3 && ['--info', '-I'].includes(args[2])) {
    native.init()
    console.log(native.partitions())
}
else if (args.length == 4 && ['--explorer', '-E'].includes(args[2])) {
    const fsPath = args[3]
    if (native.open(fsPath)) console.log(`unsupported filesystem in: ${fsPath}. Supported fs is HFS+ only`)
    else {
        native.loadFS()
        let isWorking = true
        while (isWorking) {
            const parse_cmd = (it) => { return { action: it.substring(0, 2), arg: it.substring(3) } }
            const cmd = reader.question("_> ")
            if (cmd === "pwd") console.log(native.pwd())
            else if (cmd == "exit") isWorking = false
            else {
                const { action, arg } = parse_cmd(cmd)
                console.log(arg)
                if (action === "ls") console.log(native.ls((arg === "") ? "/" : arg))
                else if (action === "cp") console.log((arg.split(" ").length !== 2) ? "you should set 2 arguments for `cp` command" : native.cp(arg))
                else if (action === "cd") console.log(native.cd(arg))
                else console.log(`supported commands:
                    pwd : show path inside mounted fs
                    ls <path> : show items in path
                    cd <path> : moves into mounted fs
                    cp <src> <dest> : copy file or folder inside mounted os
                    exit : terminate process
                `)
            }
        }
    }
} else {
    console.log("error, mode don't specified, start with -H")
    process.exit(-1)
}